plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/PostgreSqlClient.java

```diff
@@ -378,13 +380,20 @@ protected void renameTable(ConnectorSession session, Connection connection, Stri
     }
 
     @Override
-    public PreparedStatement getPreparedStatement(Connection connection, String sql)
+    public PreparedStatement getPreparedStatement(Connection connection, String sql, Optional<Integer> columnCount)
             throws SQLException
     {
         // fetch-size is ignored when connection is in auto-commit
         connection.setAutoCommit(false);
         PreparedStatement statement = connection.prepareStatement(sql);
-        statement.setFetchSize(1000);
+        if (disableAutomaticFetchSize) {
+            statement.setFetchSize(1000);
+        }
+        // This is a heuristic, not exact science. A better formula can perhaps be found with measurements.
+        // Column count is not known for non-SELECT queries. Not setting fetch size for these.
+        else if (columnCount.isPresent()) {
+            statement.setFetchSize(min(100_000 / columnCount.get(), 1_000));
+        }
```

Comment:
Perhaps, we can avoid overriding the fetch size when the size is specified with `defaultRowFetchSize` connection property in addition to this change? It will allow users configure the value in their side. We can retrieve the value with `PgConnection#getDefaultFetchSize`.

Reply:
Sounds like this is orthogonal, i.e. the existing code was setting the fetch size unconditionally, and this one sets fetch size unconditionally, just with a different value.

However, i don't think we should go into that direction at all.
if our intent is to let users configure the fetch size, we should have an explicit toggle rather than inspect fetch size provided in the JDBC URL string. Note however that giving users' control is good, but having our code be smarter is even better, and those things are at odds. i DO think we should desist desire to throw toggle at a problem.